### ISSUE TITLE
savannah: ctrl-\ hotlist key won't work under tty<i>

### ISSUE THREAD

**Issue by mc-butler** [MidnightCommander/mc#21]
> [!IMPORTANT]
> This issue was migrated from Trac:
>
> | **Origin** | https://midnight-commander.org/ticket/21 |
> | :------------- | :------------------------------------------------------------------------------------------------ |
> | **Reporter** | `ossi` (@ossilator) |
> | **Mentions** | `sbrabec@….cz`, `powerman-asdf@….ru` |

<p>
Original: <a class="ext-link" href="http://savannah.gnu.org/bugs/?13729">http://savannah.gnu.org/bugs/?13729</a><br />
</p>
<table class="wiki">
<tr><td>Submitted by:</td><td>Oswald Buddenhagen &lt;ossi&gt;</td><td>Submitted on:</td><td>Mon 11 Jul 2005 04:00:03 PM UTC
</td></tr><tr><td>Category:</td><td>Keyboard input</td><td>Severity:</td><td>3 - Normal
</td></tr><tr><td>Status:</td><td>None</td><td>Privacy:</td><td>Public
</td></tr><tr><td>Assigned to:</td><td>None</td><td>Open/Closed:</td><td>Open
</td></tr><tr><td>Release:</td><td>current (CVS or snapshot)</td><td>Operating System:</td><td>GNU/Linux
</td></tr></table>
<p>
Original submission:<br />
</p>

```
in the linux console with a german keyboard layout, one is unable 
to get to the directory hotlist via ctrl-\.
in this layout, \ is altgr-ß (equivalent to ralt-= on an american 
keyboard). i suspect that mc refuses to activate the hotkey, 
because it finds ralt to be pressed.
this behavior is new in mc 4.6. 
```
<p>
Comment 1 by Oswald Buddenhagen &lt;ossi&gt; at Wed 21 Sep 2005 05:19:41 PM UTC:<br />
</p>

```
i came up with this hack - it just makes the right alt key not be 
recognized as an alt key at all.
i have no idea how to fix it properly - one'd have to ask the 
keymap whether the ralt is part of the keycode or an additional 
modifier.
hard-coding exceptions activated after a keymap query might work, 
but is utterly ugly. well, the code already has similar hacks to 
accomodate all types of terminals ...
```
<p>
Comment 2 by Roland Eggner &lt;robiol262&gt; at Wed 29 Mar 2006 12:54:13 PM UTC:<br />
</p>

```
I found a partial solution:
Ctrl+4 (left ctrl key + "4" on main part of keyboard)
brings mc hotlist, IF running in KDE console.
Works with my Notebook Toshiba Satellite 3000-514
and this versions: SuSE-8.2/KDE-3.1.1/mc-4.6.0-336,
and also Debian-etch/KDE-3.5.1/mc-4.6.1-1.
If it works on other machines too: should be added to help files.

More satisfactory would be an additional keybinding
- I suppose ctrl+H - that works with german
keyboard layout outside of X, too.
Until now, outside of X I must use menu entry,
have no keybinding for accessing the hotlist.
Only "add to hotlist" Ctrl-X H works.
```

> [!NOTE]
> Original attachments:
>
> * [0001-add-hotkey-M-C-for-hotlist.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/21/0001-add-hotkey-M-C-for-hotlist.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/21/0001-add-hotkey-M-C-for-hotlist.patch)) by `iNode` (@inode) on <span title="2009-06-22T05:58:26+00:00">`Jun 22, 2009 at 5:58 UTC`</span>
> * [0001-make-hotkeys-which-involve-alt-gr-work-in-the-linux-.patch](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/21/0001-make-hotkeys-which-involve-alt-gr-work-in-the-linux-.patch) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/21/0001-make-hotkeys-which-involve-alt-gr-work-in-the-linux-.patch)) by `ossi` (@ossilator) on <span title="2020-10-08T19:08:26+00:00">`Oct 8, 2020 at 19:08 UTC`</span>

**Comment by zyv** [comment]
@ossilator so as a bottom line, is there a big downside to treating AltGr as Alt (as in does it break anything) or could we just commit it?

I tried to read through your argument, and apparently the first suggested approach was a hack for a particular key combination, which I think is unacceptable.

Removing the distinction between Alt and AltGr seems fine to me, but you said yourself that no one does it, and it's not to be expected, or did I read that wrong?

Is there even a "correct" solution?

I'm asking because we could

1. Commit the hack
2. Close the issue as a wontfix
3. Leave it open to implement the theoretical "correct" solution
 
and I'm not sure what that should be.

**Comment by ossilator** [comment]
no, i want to _introduce_ a distinction between alt (lalt) and altgr (ralt), because that's what the international keymaps really do, while mc incorrectly conflates the modifiers.

i wouldn't mind committing my hack, but i suspect Someone (TM) would complain, hence the suggestion to make it optional (as FAR does), or even auto-detect whether it should be enabled (by querying the keymap for ralt's meaning, using whichever "side channel" mc is using to query the modifier state in the first place (x or the vt)).

i wonder whether kitty's extended keyboard protocol takes this into consideration when reporting the alt state?

**Comment by zyv** [comment]
I think this issue might be resolved by #4769.

**Comment by ossilator** [comment]
yes, it would be. thought the question of an adequate fallback remains.
